engine/proofview.mli

```diff
@@ -208,6 +208,16 @@ val tclORELSE : 'a tactic -> (Exninfo.iexn -> 'a tactic) -> 'a tactic
     exception can be assumed non critical. *)
 val tclIFCATCH : 'a tactic -> ('a -> 'b tactic) -> (Exninfo.iexn -> 'b tactic) -> 'b tactic
 
+(** [tclWRAPFINALLY before tac finally] runs [before] before each
+    entry-point of [tac] and passes the result of [before] to
+    [finally], which is then run at each exit-point of [tac],
+    regardless of whether it succeeds or fails.  Said another way, if
+    [tac] succeeds, then it behaves as [before >>= fun v -> tac >>= fun
+    ret -> finally v <*> tclUNIT ret]; otherwise, if [tac] fails with
+    [e], it behaves as [before >>= fun v -> finally v <*> tclZERO
+    e]. *)
+val tclWRAPFINALLY : 'a tactic -> 'b tactic -> ('a -> unit tactic) -> 'b tactic
```

Comment:
A few things:
- I am not sure this is worth to export, this seems quite a specific primitive. I'd rather have it defined only in the profiling module.
- Furthermore, it doesn't rely on `Proofview` internals, so even if you want to make it public I'd rather put it in `Tacticals`.
- Also, the tactic implementation is a bit convoluted. I think it's better not to take the `before` argument and pass instead a `unit tactic` for the third argument. You can then recover the current code by extruding the bind and function application outside of the combinator.

Reply:
I suspect the documentation is lacking.
- It's exactly the same primitive needed for both the profiler and `with_strategy`
- I can move it to `Tacticals`, and could even reimplement it using `tclCASE`, if that's preferred
- I think you are mis-reading the implementation and/or the doc (and/or I got the doc wrong).  The `before` tactic is re-run before *each* entry-point/success/failure of `tac`, and the value from *that particular run* is passed to the `finally` tactic regardless of whether `tac` had another success or whether it failed.  Extruding the bind and application will not work, as the initializer will only be run once, not before every entry-point.

Reply:
I'm a bit confused where in `Tacticals` this primitive belongs.  I don't think it belongs in `New`, because the comment says
```
(** The tacticals in the module [New] are the tactical of Ltac. Their
    semantics is an extension of the tacticals in this file for the
    multi-goal backtracking tactics. They do not have the same
    semantics as the similarly named tacticals in [Proofview]. The
    tactical of [Proofview] are used in the definition of the
    tacticals of [Tacticals.New], but they are more atomic. In
    particular [Tacticals.New.tclORELSE] sees lack of progress as a
    failure, whereas [Proofview.tclORELSE] doesn't. Additionally every
    tactic which can catch failure ([tclOR], [tclORELSE], [tclTRY],
    [tclREPEAt], etc…) are run into each goal independently (failures
    and backtracks are localised to a given goal). *)
```
and it's quite important that this tactical *doesn't* localize the failures it wraps to individual goals.  On the other hand, I don't think it belongs outside of `New`, because it has the wrong type (and it very much needs to be able to talk about `'a tactic` and not just `unit tactic`)

Reply:
> It's exactly the same primitive needed for both the profiler and with_strategy

Well, in both cases this is a hack to workaround an underlying imperative issue, I don't think we should promote this as part of the standard tactic API.

> The before tactic is re-run before each entry-point/success/failure of tac, and the value from that particular run is passed to the finally tactic regardless of whether tac had another success or whether it failed

You mean that at all call sites you rely on the OCaml side-effects generated by the closure? This is unexpected and not clear from the types exported in the API. The standard way to signal this would be to take a thunk `unit -> 'b tactic` instead of a function `'a -> 'b tactic`. But anyways, I'd really prefer to have this defined where it is used.

Reply:
No, I do not rely on side effects of the closure.  Consider the case of the profiler: the `before` tactic, when run, gets the current time (lifted into the nonlogical tactic monad, so the time corresponds to when the tactic is run, not evaluated). The `after` tactic, when run, updates the profiling info with the time elapsed in the current tactic.  This again happens when running the tactic, not when evaluating the OCaml thunk.  But it's important that the `finally` tactic get a *different* start time for each success (or failure).  Namely, it should be getting the start time of that particular run, not the time the tactic was initially entered into.  (Consider `tac1 := slow1 + slow2` and then profiling `try (tac1; slow3; fail)`.  When we re-enter `tac1` to execute `slow2` after failing, we need a new start time that doesn't include the time we just spent in `slow3`.)

My concern about defining this at the use-site is that it's complicated enough that I'd rather not define it in two places (the profiler and with_strategy) and I'm not sure how else to not duplicate code.  (And I do also think it's a genuinely sensible primitive of the backtracking monad, allowing proper handling of stateful side-effects not tracked by the monad itself.)

Reply:
I've moved this to be local to the Ltac Profiler